inference-engine/thirdparty/clDNN/src/gpu/loop_gpu.cpp

```diff
@@ -0,0 +1,399 @@
+                }
+            }
+
+            // Set backedges
+            for (const auto& backedge_memory_mapping : instance.backedge_memory_mappings) {
+                backedge_memory_mapping.setup_iteration(current_iteration);
+            }
+
+            // Set sliced output memory
+            for (const auto& concat_output_mem_mapping : concatenated_output_mem_mappings) {
+                concat_output_mem_mapping.setup_concatenated_output_memory(current_iteration);
+            }
+            if (current_iteration == 0) {
+                body_network->execute(events);
+            } else {
+                body_network->execute(loop_carried_dep);
+            }
+            loop_carried_dep.clear();
+            for (const auto& backedge : node.get_back_edges()) {
+                event_impl::ptr body_event = body_network->get_primitive_event(backedge.from);
+                loop_carried_dep.emplace_back(body_event);
+            }
+
+            //TODO: "curreint_iteration primitive and execution_condition is prepared
+            //as they are presented in the ngraph opset document for loop operation.
+            //However they are not being used yet and only TensorIterator which has fixed sequence length is being validated.
+            if (node.is_current_iteration_used()) {
+                write_scalar_value(*current_iteration_mem, current_iteration);
+            }
+            if (node.is_execution_condition_used()) {
+                execution_condition = read_scalar_value(*execution_condition_mem);
+            }
+            // update index & execution condition for the next iteration
+            ++current_iteration;
+        }
+
+        body_network->reset_execution();
+
+        // Concatenate sliced output to the outer network
+        for (size_t i = 0; i < concatenated_output_mem_mappings.size(); ++i) {
+            const auto& concat_output = concatenated_output_mem_mappings.at(i);
+            concat_output.restore_concatenated_mem();
+        }
+
+        const primitive_id& num_iteration_id = node.get_num_iteration_id();
+        memory_impl& num_iteration_mem = outer_network.get_primitive(num_iteration_id)->output_memory();
+        write_scalar_value(num_iteration_mem, current_iteration);
+
+        dynamic_cast<cldnn::user_event*>(ev.get())->set();
+        return ev;
+    }
+
+    static primitive_impl* create(const loop_node& arg) { return new loop_gpu(arg); }
+};
+
+namespace detail {
+attach_loop_gpu::attach_loop_gpu() {
+    std::vector<data_types> loop_data_types{ data_types::bin, data_types::u8, data_types::i8, data_types::f16,
+                                             data_types::f32, data_types::i32, data_types::i64};
+
+    std::vector<format> loop_formats{ format::bfyx, format::bfzyx, format::bfwzyx };
+
+    for (const data_types loop_data_type : loop_data_types) {
+        for (const format loop_format : loop_formats) {
+            implementation_map<loop>::add(
```

Comment:
If this primitive doesn't depend on precision/layout, then you can add specialization of implementation_key template and change key type. See reorder primitive as an example.

Reply:
I added template specialization of `implementation_key` and will be applied in the next commit.

Reply:
Fixed